Add script/string roundtrip fuzz tests

sipa opened the pull request:
Builds on top of #105.

This adds script and string roundtrip fuzz tests, all in the same source file (reusing some logic), and unified the names of the tests:
* miniscript_stable (the old miniscript_random)
* miniscript_smart (the old miniscript_random_smart)
* miniscript_script (round-trip from/to script)
* miniscript_string (round-trip from/to string)

darosior commented on bitcoin/test/fuzz/miniscript_random.cpp:
```diff
@@ -606,7 +592,7 @@ void TestNode(const NodeRef& node, FuzzedDataProvider& provider)
 
 /** Fuzz target that runs TestNode on nodes generated using ConsumeNodeStable. */
-FUZZ_TARGET_INIT(miniscript_random_stable, initialize_miniscript_random)
+FUZZ_TARGET(miniscript_random_stable)
```
Neat.

darosior commented on bitcoin/test/fuzz/miniscript_random.cpp:
```diff
@@ -639,3 +676,16 @@ FUZZ_TARGET(miniscript_string)
 }
+
+/* Fuzz tests that test parsing from a script, and roundtripping via script. */
+FUZZ_TARGET(miniscript_script)
```
This effectively replaces the `miniscript_decode` target we already have?

darosior reviewed (commented):
Code looks good to me, this is much cleaner and probably much more efficient than my previous targets. I did not test them to compare the coverage though.

I think we should get rid of my `miniscript_decode` if we have `miniscript_script`.

darosior commented on bitcoin/test/fuzz/miniscript_random.cpp:
```diff
@@ -849,3 +849,18 @@ FUZZ_TARGET(miniscript_random_smart)
+
+    std::string str2;
+    assert(parsed->ToString(PARSER_CTX, str2));
+    auto parsed2 = miniscript::FromString(str, PARSER_CTX);
```
I think you meant to parse from `str2` here?

sipa replied:
Done.

darosior reviewed (commented):
Apart from the typo in the minscript_string target mentioned above, this looks good to me.

darosior commented on bitcoin/test/fuzz/miniscript_random.cpp:
```diff
@@ -67,12 +68,11 @@ struct TestData {
         }
     }
-};
+} TEST_DATA;
```
Actually we can't do that, it's causing a fuzz failure when running other targets which `ECC_Start()` too. I think it's because all targets are linked into a single binary, and that it gets executed twice for those that already call it during their initialization?

MarcoFalke replied:
Apart from the crash, globals will also poison the coverage data, as unrelated test cases will mark the constructor of the global as covered.

Ideally each target runs only its own code. Expensive setup code can be put into the init function, which is called once before the target starts.

darosior replied:
Hmm. I was thinking of just `ECC_Stop`ing at the end of the constructor (since we only need the context for the GetPubkey() there). But then maybe it's worth just not including this change and stay with my previous hack (which does exactly what Marco says)?

sipa replied:
Done.

sipa commented:
Rebased on top of merged #105, and also addressed:
* Partially reverted to FUZZ_TARGET_INIT again, to make sure initializations aren't executed when not needed.
* s/str/str2/ in the string test

darosior commented on bitcoin/test/fuzz/miniscript_random.cpp:
```diff
@@ -887,3 +924,16 @@ FUZZ_TARGET_INIT(miniscript_string, FuzzInit)
 }
+
+/* Fuzz tests that test parsing from a script, and roundtripping via script. */
+FUZZ_TARGET_INIT(miniscript_script, FuzzInit)
```
nit: I don't think that the `miniscript_script` target needs `FuzzInit` since it doesn't use anything from TEST_DATA.

sipa replied:
You're right, fixed.

darosior reviewed (commented):
ACK ef71d2c -- just a nit on `FuzzInit` not being necessary for one of the targets but no big deal.
